include/sys/dbuf.h

```diff
@@ -321,12 +321,18 @@ typedef struct dmu_buf_impl {
 	uint8_t db_dirtycnt;
 } dmu_buf_impl_t;
 
-#define	DBUF_RWLOCKS 8192
-#define	DBUF_HASH_RWLOCK(h, idx) (&(h)->hash_rwlocks[(idx) & (DBUF_RWLOCKS-1)])
+typedef struct dbuf_hash_mutex {
+	kmutex_t dhm_mutex ____cacheline_aligned;
```

Comment:
I am not sure if `____cacheline_aligned` is going to do anything since this structure is dynamically allocated. The choice of starting alignment is left to the dynamic allocator, which has no knowledge of our needs.

I am also not sure if this will add padding to bring struct members up to the size of a cacheline. If it does, then it is still useful. If it does not, then that is a pre-existing issue.

Reply:
I think ____cacheline_aligned here will pad the structure size to the cache line.  But I don't think it is useful in this case.  It allows to avoid false sharing, indeed, but since we have unlimited number of locks, we can achieve the same result just by increasing number of locks, if needed, also reducing true sharing.

I think we should drop dbuf_hash_mutex_t declaration completely and use kmutex_t directly.

Reply:
Right, the intention here was to preserve the existing layout and pad out the structure to a full cache line.  Let me do some testing with, and without, the extra padding to see if there's any measurable benefit.  If there's not, I'm all for dropping the `dbuf_hash_mutex_t` struct.

Reply:
In existing code I think ____cacheline_aligned should pad only the whole array in general. So preserving behavior would mean removing ____cacheline_aligned completely, relying on vmem_zalloc().

Reply:
Ideally, we should want the array members to be padded. Otherwise, we will end up with entries spanning two cache lines.

Reply:
On FreeBSD all `struct sx`, `struct rwlock` and `struct mutex` are all 32-byte wide, so pack perfectly.

Reply:
In that case, we could limit the padding to Linux.

Reply:
When testing locally I wasn't able to observe any measurable performance distance with or without the padding.  I'll go ahead and update this PR to drop the `dbuf_hash_mutex_t` struct and padding entirely.